File: src/pmem.c
// Copyright (c) 2026 渟雲. All rights reserved.
#include "./common.h"

// Mask for the physical frame bits of a page table entry.
#define PMEM_PMASK 0x000FFFFFFFFFF000ULL
#define PMEM_PAGE_OFFSET_SIZE 12

#define PMEM_MAX_CPU_PAGES 64

extern ULONG g_UserDirectoryTableBaseOffset;

typedef struct _PMEM_PAGE {
  PVOID VirtualAddress;
  UINT64* PteLong;
} PMEM_PAGE;

static PMEM_PAGE g_PmemPages[PMEM_MAX_CPU_PAGES];

static PVOID PmemPhysToVa(UINT64 PhysicalAddress) {
  PHYSICAL_ADDRESS phys;
  phys.QuadPart = (LONGLONG)PhysicalAddress;
  return MmGetVirtualForPhysicalTrick(phys);
}

static UINT64* PmemGetPte(UINT64 address) {
  const UINT64 cr3 = __readcr3() & ~0xFULL;

  const UINT64 pml4Index = (address >> 39) & 0x1FF;
  const UINT64 pdptIndex = (address >> 30) & 0x1FF;
  const UINT64 pdIndex = (address >> 21) & 0x1FF;
  const UINT64 ptIndex = (address >> 12) & 0x1FF;

  PVOID pml4Va = PmemPhysToVa(cr3);
  if (pml4Va == NULL) return NULL;
  const UINT64 pml4e = ((volatile UINT64*)pml4Va)[pml4Index];
  if (!(pml4e & 1)) return NULL;

  PVOID pdptVa = PmemPhysToVa(pml4e & PMEM_PMASK);
  if (pdptVa == NULL) return NULL;
  const UINT64 pdpte = ((volatile UINT64*)pdptVa)[pdptIndex];
  if (!(pdpte & 1)) return NULL;
  if (pdpte & 0x80) return NULL;  // 1GB large page, no PTE to swap

  PVOID pdVa = PmemPhysToVa(pdpte & PMEM_PMASK);
  if (pdVa == NULL) return NULL;
  const UINT64 pde = ((volatile UINT64*)pdVa)[pdIndex];
  if (!(pde & 1)) return NULL;
  if (pde & 0x80) return NULL;  // 2MB large page, no PTE to swap

  PVOID ptVa = PmemPhysToVa(pde & PMEM_PMASK);
  if (ptVa == NULL) return NULL;
  UINT64* pte = &((UINT64*)ptVa)[ptIndex];
  if (!(*pte & 1)) return NULL;
  return pte;
}

static NTSTATUS PmemSwapPhys(UINT64 PhysicalAddress, PVOID Buffer, SIZE_T Size,
                             BOOLEAN Write) {
  if (Size == 0) return STATUS_SUCCESS;

  UINT32 idx = KeGetCurrentProcessorIndex();
  if (idx >= PMEM_MAX_CPU_PAGES) idx = PMEM_MAX_CPU_PAGES - 1;
  PMEM_PAGE* page = &g_PmemPages[idx];
  if (page->VirtualAddress == NULL || page->PteLong == NULL) {
    return STATUS_UNSUCCESSFUL;
  }

  const UINT64 pageStart = PhysicalAddress & ~(PMEM_PAGE_SIZE - 1);
  const SIZE_T pageOffset = (SIZE_T)(PhysicalAddress & (PMEM_PAGE_SIZE - 1));
  if (pageOffset + Size > PMEM_PAGE_SIZE) return STATUS_INVALID_PARAMETER;

  const UINT64 oldValue = *page->PteLong;

  *page->PteLong = (oldValue & ~PMEM_PMASK) | pageStart;
  _mm_mfence();
  __invlpg(page->VirtualAddress);

  UINT8* target = (UINT8*)page->VirtualAddress + pageOffset;
  if (Write) {
    kmemmove(target, Buffer, Size);
  } else {
    kmemmove(Buffer, target, Size);
  }

  *page->PteLong = oldValue;
  _mm_mfence();
  __invlpg(page->VirtualAddress);

  return STATUS_SUCCESS;
}

NTSTATUS ReadPhysical(UINT64 PhysicalAddress, PVOID Buffer, SIZE_T Size,
                      PSIZE_T BytesRead) {
  NTSTATUS status = PmemSwapPhys(PhysicalAddress, Buffer, Size, FALSE);
  if (NT_SUCCESS(status) && BytesRead != NULL) *BytesRead = Size;
  return status;
}

NTSTATUS WritePhysical(UINT64 PhysicalAddress, const void* Buffer, SIZE_T Size) {
  return PmemSwapPhys(PhysicalAddress, (PVOID)Buffer, Size, TRUE);
}

static UINT64 ReadPhysicalU64(UINT64 PhysicalAddress) {
  UINT64 value = 0;
  NTSTATUS status = PmemSwapPhys(PhysicalAddress, &value, sizeof(value), FALSE);
  if (!NT_SUCCESS(status)) return 0;
  return value;
}

NTSTATUS InitPmemPages(VOID) {
  for (UINT32 i = 0; i < PMEM_MAX_CPU_PAGES; i++) {
    PHYSICAL_ADDRESS maxAddress;
    maxAddress.QuadPart = MAXULONG64;
    PVOID va = _MmAllocateContiguousMemory(PMEM_PAGE_SIZE, maxAddress);
    if (va == NULL) {
      CleanupPmemPages();
      return STATUS_INSUFFICIENT_RESOURCES;
    }
    UINT64* pte = PmemGetPte((UINT64)va);
    if (pte == NULL) {
      _MmFreeContiguousMemory(va);
      CleanupPmemPages();
      return STATUS_UNSUCCESSFUL;
    }
    g_PmemPages[i].VirtualAddress = va;
    g_PmemPages[i].PteLong = pte;
  }
  return STATUS_SUCCESS;
}

VOID CleanupPmemPages(VOID) {
  for (UINT32 i = 0; i < PMEM_MAX_CPU_PAGES; i++) {
    if (g_PmemPages[i].VirtualAddress != NULL) {
      _MmFreeContiguousMemory(g_PmemPages[i].VirtualAddress);
      g_PmemPages[i].VirtualAddress = NULL;
      g_PmemPages[i].PteLong = NULL;
    }
  }
}

UINT64 TranslateLinearAddress(UINT64 DirectoryTableBase,
                              UINT64 VirtualAddress) {
  DirectoryTableBase &= ~0xFULL;
  UINT64 pageOffset = VirtualAddress & ~(~0ULL << PMEM_PAGE_OFFSET_SIZE);
  UINT64 pte = (VirtualAddress >> 12) & 0x1FF;
  UINT64 pt = (VirtualAddress >> 21) & 0x1FF;
  UINT64 pd = (VirtualAddress >> 30) & 0x1FF;
  UINT64 pdp = (VirtualAddress >> 39) & 0x1FF;

  UINT64 pdpValue = ReadPhysicalU64(DirectoryTableBase + pdp * 0x08);
  if (!(pdpValue & 0x01)) return 0;

  UINT64 pdValue = ReadPhysicalU64((pdpValue & PMEM_PMASK) + pd * 0x08);
  if (!(pdValue & 0x01)) return 0;
  if (pdValue & 0x80) {
    /* 1GB large page, use pde's 12-34 bits */
    return (pdValue & (0xFFFFFFFFFFFFFFFFULL << 42 >> 12)) +
           (VirtualAddress & ~(~0ULL << 30));
  }

  UINT64 ptValue = ReadPhysicalU64((pdValue & PMEM_PMASK) + pt * 0x08);
  if (!(ptValue & 0x01)) return 0;
  if (ptValue & 0x80) {
    /* 2MB large page */
    return (ptValue & PMEM_PMASK) + (VirtualAddress & ~(~0ULL << 21));
  }

  UINT64 pteValue =
      ReadPhysicalU64((ptValue & PMEM_PMASK) + pte * 0x08) & PMEM_PMASK;
  if (pteValue == 0) return 0;
  return pteValue + pageOffset;
}

UINT64 GetProcessCr3(PEPROCESS Process) {
  if (Process == NULL) return 0;
  // KPROCESS.DirectoryTableBase is at offset 0x28 on x64.
  UINT64 dtb = *(volatile UINT64*)((UINT_PTR)Process + 0x28);
  if (dtb != 0) return dtb;
  // Fall back to UserDirectoryTableBase (KPTI user-mode CR3).
  if (g_UserDirectoryTableBaseOffset == 0) {
    if (!InitOffsetsByVersion()) return 0;
  }
  return *(volatile UINT64*)((UINT_PTR)Process +
                             g_UserDirectoryTableBaseOffset);
}

NTSTATUS ReadProcessMemory(PEPROCESS Process, UINT64 VirtualAddress,
                           PVOID Buffer, SIZE_T Size) {
  if (Process == NULL || Buffer == NULL) return STATUS_INVALID_PARAMETER;
  UINT64 dtb = GetProcessCr3(Process);
  if (dtb == 0) return STATUS_INVALID_PARAMETER;

  UINT8* dst = (UINT8*)Buffer;
  SIZE_T offset = 0;
  while (offset < Size) {
    SIZE_T remaining = Size - offset;
    UINT64 va = VirtualAddress + offset;
    SIZE_T chunk = PMEM_PAGE_SIZE - (SIZE_T)(va & (PMEM_PAGE_SIZE - 1));
    if (chunk > remaining) chunk = remaining;

    UINT64 phys = TranslateLinearAddress(dtb, va);
    if (phys == 0) return STATUS_PARTIAL_COPY;

    SIZE_T done = 0;
    NTSTATUS status = ReadPhysical(phys, dst + offset, chunk, &done);
    if (!NT_SUCCESS(status)) return status;
    offset += chunk;
  }
  return STATUS_SUCCESS;
}

NTSTATUS WriteProcessMemory(PEPROCESS Process, UINT64 VirtualAddress,
                            const void* Buffer, SIZE_T Size) {
  if (Process == NULL || Buffer == NULL) return STATUS_INVALID_PARAMETER;
  UINT64 dtb = GetProcessCr3(Process);
  if (dtb == 0) return STATUS_INVALID_PARAMETER;

  const UINT8* src = (const UINT8*)Buffer;
  SIZE_T offset = 0;
  while (offset < Size) {
    SIZE_T remaining = Size - offset;
    UINT64 va = VirtualAddress + offset;
    SIZE_T chunk = PMEM_PAGE_SIZE - (SIZE_T)(va & (PMEM_PAGE_SIZE - 1));
    if (chunk > remaining) chunk = remaining;

    UINT64 phys = TranslateLinearAddress(dtb, va);
    if (phys == 0) return STATUS_PARTIAL_COPY;

    NTSTATUS status = WritePhysical(phys, src + offset, chunk);
    if (!NT_SUCCESS(status)) return status;
    offset += chunk;
  }
  return STATUS_SUCCESS;
}

NTSTATUS CopyVirtualMemory(PEPROCESS FromProcess, UINT64 FromAddress,
                           PEPROCESS ToProcess, UINT64 ToAddress, SIZE_T Size) {
  if (FromProcess == NULL || ToProcess == NULL || Size == 0)
    return STATUS_INVALID_PARAMETER;

  UINT64 fromDtb = GetProcessCr3(FromProcess);
  UINT64 toDtb = GetProcessCr3(ToProcess);
  if (fromDtb == 0 || toDtb == 0) return STATUS_INVALID_PARAMETER;

  UINT8* scratch =
      (UINT8*)_ExAllocatePool2(POOL_FLAG_NON_PAGED | POOL_FLAG_UNINITIALIZED,
                               PMEM_PAGE_SIZE, 0x446C6148);
  if (scratch == NULL) {
    return STATUS_INSUFFICIENT_RESOURCES;
  }

  NTSTATUS status = STATUS_SUCCESS;
  SIZE_T offset = 0;

  while (offset < Size) {
    SIZE_T remaining = Size - offset;
    UINT64 fva = FromAddress + offset;
    UINT64 tva = ToAddress + offset;

    SIZE_T chunk = PMEM_PAGE_SIZE - (SIZE_T)(fva & (PMEM_PAGE_SIZE - 1));
    SIZE_T chunk2 = PMEM_PAGE_SIZE - (SIZE_T)(tva & (PMEM_PAGE_SIZE - 1));
    if (chunk2 < chunk) chunk = chunk2;
    if (chunk > remaining) chunk = remaining;

    UINT64 fphys = TranslateLinearAddress(fromDtb, fva);
    if (fphys == 0) {
      status = STATUS_PARTIAL_COPY;
      goto Cleanup;
    }

    SIZE_T done = 0;
    status = ReadPhysical(fphys, scratch, chunk, &done);
    if (!NT_SUCCESS(status)) {
      goto Cleanup;
    }

    UINT64 tphys = TranslateLinearAddress(toDtb, tva);
    if (tphys == 0) {
      status = STATUS_PARTIAL_COPY;
      goto Cleanup;
    }

    status = WritePhysical(tphys, scratch, chunk);
    if (!NT_SUCCESS(status)) {
      goto Cleanup;
    }

    offset += chunk;
  }

Cleanup:
  RtlSecureZeroMemory(scratch, PMEM_PAGE_SIZE);
  _ExFreePoolWithTag(scratch, 0x446C6148);

  return status;
}

File: src/pmem.rs
// Copyright (c) 2026 渟雲. All rights reserved.

use core::ffi::c_void;
use core::ptr::null_mut;

use crate::consts::*;
use crate::ffi::*;
use crate::imports::{
    _ExAllocatePool2, _ExFreePoolWithTag, _MmAllocateContiguousMemory, _MmFreeContiguousMemory,
};
use crate::process::{G_USER_DIRECTORY_TABLE_BASE_OFFSET, init_offsets_by_version};
use crate::types::NtStatus;

type FnMmAllocateContiguousMemory = unsafe extern "system" fn(usize, u64) -> *mut c_void;
type FnMmFreeContiguousMemory = unsafe extern "system" fn(*mut c_void);
type FnExAllocatePool2 = unsafe extern "system" fn(u64, usize, u32) -> *mut c_void;
type FnExFreePoolWithTag = unsafe extern "system" fn(*mut c_void, u32);

#[derive(Clone, Copy)]
struct PMemPage {
    virtual_address: *mut c_void,
    pte_long: *mut u64,
}

static mut G_PMEM_PAGES: [PMemPage; PMEM_MAX_CPU_PAGES] = [PMemPage {
    virtual_address: null_mut(),
    pte_long: null_mut(),
}; PMEM_MAX_CPU_PAGES];

fn read_cr3() -> u64 {
    let cr3: u64;
    unsafe {
        core::arch::asm!("mov {}, cr3", out(reg) cr3, options(nostack));
    }
    cr3
}

fn mfence() {
    unsafe {
        core::arch::asm!("mfence", options(nostack));
    }
}

fn invlpg(address: u64) {
    unsafe {
        core::arch::asm!("invlpg [{0}]", in(reg) address, options(nostack));
    }
}

fn current_processor_index() -> u32 {
    let idx: u32;
    unsafe {
        core::arch::asm!("mov eax, gs:[0x1C]", out("eax") idx, options(nostack));
    }
    idx
}

fn secure_zero(ptr: *mut u8, len: usize) {
    unsafe {
        core::ptr::write_bytes(ptr, 0, len);
    }
}

fn pmem_phys_to_va(physical_address: u64) -> *mut c_void {
    crate::reimpl::mm_get_virtual_for_physical_trick(physical_address)
}

fn pmem_get_pte(address: u64) -> *mut u64 {
    unsafe {
        let cr3 = read_cr3() & !0xF;
        let pml4_index = (address >> 39) & 0x1FF;
        let pdpt_index = (address >> 30) & 0x1FF;
        let pd_index = (address >> 21) & 0x1FF;
        let pt_index = (address >> 12) & 0x1FF;

        let pml4_va = pmem_phys_to_va(cr3);
        if pml4_va.is_null() {
            return null_mut();
        }
        let pml4e = *(pml4_va as *const u64).add(pml4_index as usize);
        if pml4e & 1 == 0 {
            return null_mut();
        }

        let pdpt_va = pmem_phys_to_va(pml4e & PMEM_PMASK);
        if pdpt_va.is_null() {
            return null_mut();
        }
        let pdpte = *(pdpt_va as *const u64).add(pdpt_index as usize);
        if pdpte & 1 == 0 {
            return null_mut();
        }
        if pdpte & 0x80 != 0 {
            return null_mut();
        }

        let pd_va = pmem_phys_to_va(pdpte & PMEM_PMASK);
        if pd_va.is_null() {
            return null_mut();
        }
        let pde = *(pd_va as *const u64).add(pd_index as usize);
        if pde & 1 == 0 {
            return null_mut();
        }
        if pde & 0x80 != 0 {
            return null_mut();
        }

        let pt_va = pmem_phys_to_va(pde & PMEM_PMASK);
        if pt_va.is_null() {
            return null_mut();
        }
        let pte = (pt_va as *mut u64).add(pt_index as usize);
        if *pte & 1 == 0 {
            return null_mut();
        }
        pte
    }
}

fn pmem_swap_phys(
    physical_address: u64,
    buffer: *mut c_void,
    size: usize,
    write: bool,
) -> NtStatus {
    unsafe {
        if size == 0 {
            return STATUS_SUCCESS;
        }

        let mut idx = current_processor_index();
        if idx as usize >= PMEM_MAX_CPU_PAGES {
            idx = (PMEM_MAX_CPU_PAGES - 1) as u32;
        }
        let page = &mut G_PMEM_PAGES[idx as usize];
        if page.virtual_address.is_null() || page.pte_long.is_null() {
            return STATUS_UNSUCCESSFUL;
        }

        let page_start = physical_address & !(PMEM_PAGE_SIZE as u64 - 1);
        let page_offset = (physical_address & (PMEM_PAGE_SIZE as u64 - 1)) as usize;
        if page_offset + size > PMEM_PAGE_SIZE {
            return STATUS_INVALID_PARAMETER;
        }

        let old_value = *page.pte_long;
        *page.pte_long = (old_value & !PMEM_PMASK) | page_start;
        mfence();
        invlpg(page.virtual_address as u64);

        let target = (page.virtual_address as *mut u8).add(page_offset);
        if write {
            kmemmove(target as *mut c_void, buffer, size);
        } else {
            kmemmove(buffer, target as *const c_void, size);
        }

        *page.pte_long = old_value;
        mfence();
        invlpg(page.virtual_address as u64);

        STATUS_SUCCESS
    }
}

pub fn read_physical(physical_address: u64, buffer: *mut c_void, size: usize) -> NtStatus {
    let status = pmem_swap_phys(physical_address, buffer, size, false);
    status
}

pub fn write_physical(physical_address: u64, buffer: *const c_void, size: usize) -> NtStatus {
    pmem_swap_phys(physical_address, buffer as *mut c_void, size, true)
}

fn read_physical_u64(physical_address: u64) -> u64 {
    let mut value = 0u64;
    let status = pmem_swap_phys(
        physical_address,
        (&mut value) as *mut u64 as *mut c_void,
        8,
        false,
    );
    if status < 0 {
        return 0;
    }
    value
}

pub fn init_pmem_pages() -> NtStatus {
    unsafe {
        for i in 0..PMEM_MAX_CPU_PAGES {
            let va = if _MmAllocateContiguousMemory.is_null() {
                null_mut()
            } else {
                let f: FnMmAllocateContiguousMemory =
                    core::mem::transmute(_MmAllocateContiguousMemory);
                f(PMEM_PAGE_SIZE, u64::MAX)
            };
            if va.is_null() {
                cleanup_pmem_pages();
                return STATUS_INSUFFICIENT_RESOURCES;
            }
            let pte = pmem_get_pte(va as u64);
            if pte.is_null() {
                if !_MmFreeContiguousMemory.is_null() {
                    let f: FnMmFreeContiguousMemory = core::mem::transmute(_MmFreeContiguousMemory);
                    f(va);
                }
                cleanup_pmem_pages();
                return STATUS_UNSUCCESSFUL;
            }
            G_PMEM_PAGES[i].virtual_address = va;
            G_PMEM_PAGES[i].pte_long = pte;
        }
        STATUS_SUCCESS
    }
}

pub fn cleanup_pmem_pages() {
    unsafe {
        for i in 0..PMEM_MAX_CPU_PAGES {
            if !G_PMEM_PAGES[i].virtual_address.is_null() {
                if !_MmFreeContiguousMemory.is_null() {
                    let f: FnMmFreeContiguousMemory = core::mem::transmute(_MmFreeContiguousMemory);
                    f(G_PMEM_PAGES[i].virtual_address);
                }
                G_PMEM_PAGES[i].virtual_address = null_mut();
                G_PMEM_PAGES[i].pte_long = null_mut();
            }
        }
    }
}

pub fn translate_linear_address(directory_table_base: u64, virtual_address: u64) -> u64 {
    let directory_table_base = directory_table_base & !0xF;
    let page_offset = virtual_address & !(!0u64 << 12);
    let pte = (virtual_address >> 12) & 0x1FF;
    let pt = (virtual_address >> 21) & 0x1FF;
    let pd = (virtual_address >> 30) & 0x1FF;
    let pdp = (virtual_address >> 39) & 0x1FF;

    let pdp_value = read_physical_u64(directory_table_base + pdp * 0x08);
    if pdp_value & 0x01 == 0 {
        return 0;
    }

    let pd_value = read_physical_u64((pdp_value & PMEM_PMASK) + pd * 0x08);
    if pd_value & 0x01 == 0 {
        return 0;
    }
    if pd_value & 0x80 != 0 {
        return (pd_value & (0xFFFF_FFFF_FFFF_FFFFu64 << 42 >> 12))
            + (virtual_address & !(!0u64 << 30));
    }

    let pt_value = read_physical_u64((pd_value & PMEM_PMASK) + pt * 0x08);
    if pt_value & 0x01 == 0 {
        return 0;
    }
    if pt_value & 0x80 != 0 {
        return (pt_value & PMEM_PMASK) + (virtual_address & !(!0u64 << 21));
    }

    let pte_value = read_physical_u64((pt_value & PMEM_PMASK) + pte * 0x08) & PMEM_PMASK;
    if pte_value == 0 {
        return 0;
    }
    pte_value + page_offset
}

pub fn get_process_cr3(process: *mut c_void) -> u64 {
    if process.is_null() {
        return 0;
    }
    unsafe {
        let dtb = *(process as *const u64).add(0x28 / 8);
        if dtb != 0 {
            return dtb;
        }
        if G_USER_DIRECTORY_TABLE_BASE_OFFSET == 0 {
            if !init_offsets_by_version() {
                return 0;
            }
        }
        *(process as *const u64).add(G_USER_DIRECTORY_TABLE_BASE_OFFSET as usize / 8)
    }
}

pub fn read_process_memory(
    process: *mut c_void,
    virtual_address: u64,
    buffer: *mut c_void,
    size: usize,
) -> NtStatus {
    unsafe {
        if process.is_null() || buffer.is_null() {
            return STATUS_INVALID_PARAMETER;
        }
        let dtb = get_process_cr3(process);
        if dtb == 0 {
            return STATUS_INVALID_PARAMETER;
        }

        let irql = KzRaiseIrqlMeme(DISPATCH_LEVEL);
        let dst = buffer as *mut u8;
        let mut offset = 0usize;
        let mut status = STATUS_SUCCESS;
        while offset < size {
            let remaining = size - offset;
            let va = virtual_address + offset as u64;
            let mut chunk = PMEM_PAGE_SIZE - (va & (PMEM_PAGE_SIZE as u64 - 1)) as usize;
            if chunk > remaining {
                chunk = remaining;
            }

            let phys = translate_linear_address(dtb, va);
            if phys == 0 {
                status = STATUS_PARTIAL_COPY;
                break;
            }

            let rstatus = read_physical(phys, dst.add(offset) as *mut c_void, chunk);
            if rstatus < 0 {
                status = rstatus;
                break;
            }
            offset += chunk;
        }
        KzLowerIrqlMeme(irql);
        status
    }
}

pub fn copy_virtual_memory(
    from_process: *mut c_void,
    from_address: u64,
    to_process: *mut c_void,
    to_address: u64,
    size: usize,
) -> NtStatus {
    unsafe {
        if from_process.is_null() || to_process.is_null() || size == 0 {
            return STATUS_INVALID_PARAMETER;
        }

        let from_dtb = get_process_cr3(from_process);
        let to_dtb = get_process_cr3(to_process);
        if from_dtb == 0 || to_dtb == 0 {
            return STATUS_INVALID_PARAMETER;
        }

        let scratch = if _ExAllocatePool2.is_null() {
            null_mut()
        } else {
            let f: FnExAllocatePool2 = core::mem::transmute(_ExAllocatePool2);
            f(
                POOL_FLAG_NON_PAGED | POOL_FLAG_UNINITIALIZED,
                PMEM_PAGE_SIZE,
                POOL_TAG_COPY,
            )
        };
        if scratch.is_null() {
            return STATUS_INSUFFICIENT_RESOURCES;
        }

        let mut status = STATUS_SUCCESS;
        let mut offset = 0usize;
        while offset < size {
            let remaining = size - offset;
            let fva = from_address + offset as u64;
            let tva = to_address + offset as u64;

            let mut chunk = PMEM_PAGE_SIZE - (fva & (PMEM_PAGE_SIZE as u64 - 1)) as usize;
            let chunk2 = PMEM_PAGE_SIZE - (tva & (PMEM_PAGE_SIZE as u64 - 1)) as usize;
            if chunk2 < chunk {
                chunk = chunk2;
            }
            if chunk > remaining {
                chunk = remaining;
            }

            let fphys = translate_linear_address(from_dtb, fva);
            if fphys == 0 {
                status = STATUS_PARTIAL_COPY;
                break;
            }

            let rstatus = read_physical(fphys, scratch, chunk);
            if rstatus < 0 {
                status = rstatus;
                break;
            }

            let tphys = translate_linear_address(to_dtb, tva);
            if tphys == 0 {
                status = STATUS_PARTIAL_COPY;
                break;
            }

            let wstatus = write_physical(tphys, scratch, chunk);
            if wstatus < 0 {
                status = wstatus;
                break;
            }

            offset += chunk;
        }

        secure_zero(scratch as *mut u8, PMEM_PAGE_SIZE);
        if !_ExFreePoolWithTag.is_null() {
            let f: FnExFreePoolWithTag = core::mem::transmute(_ExFreePoolWithTag);
            f(scratch, POOL_TAG_COPY);
        }

        status
    }
}
